site/source/tutorials/1015_form_validation/index.md

```diff
@@ -0,0 +1,366 @@
+		id: this._errorId,
+		classes: this.classes(css.error),
+		'aria-live': 'polite'
+	}, [ 'Please enter valid text for this field' ]) : null
+])
+```
+
+The error message is associated with the text input through `aria-describedby`, and the `aria-live` attribute ensures it will be read if it is added to the DOM or changed. Wrapping both the input and the error message in a containing `<div>` allows us to position the error message relative to the input if desired.
+
+{% instruction 'Extend `TextInput` to create a `ValidatedTextInput` widget with an error message and `onValidate` method' %}
+
+Re-creating the same error message boilerplate for multiple text inputs seems overly repetitive, so we're going to extend `TextInput` instead. This will also allow us to have better control over when validation occurs, e.g. by adding it to blur events as well. For now, just create a `ValidatedTextInput` widget that accepts the same properties interface as `TextInput` but with an `errorMessage` string and `onValidate` method. It should return the same node structure modeled above.
+
+{% solution showsolution4 %}
+```ts
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { TypedTargetEvent } from '@dojo/widget-core/interfaces';
+import { v, w } from '@dojo/widget-core/d';
+import uuid from '@dojo/core/uuid';
+import { ThemeableMixin, theme } from '@dojo/widget-core/mixins/Themeable';
+import TextInput, { TextInputProperties } from '@dojo/widgets/textinput/TextInput';
+import * as css from '../styles/workerForm.css';
+
+export interface ValidatedTextInputProperties extends TextInputProperties {
+	errorMessage?: string;
+	onValidate?: (event: Event) => void;
+}
+
+export const ValidatedTextInputBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class ValidatedTextInput extends ValidatedTextInputBase<ValidatedTextInputProperties> {
+	private _errorId = uuid();
+
+	protected render() {
+		const {
+			disabled,
+			label,
+			maxLength,
+			minLength,
+			name,
+			placeholder,
+			readOnly,
+			required,
+			type = 'text',
+			value,
+			invalid,
+			errorMessage,
+			onBlur,
+			onInput
+		} = this.properties;
+
+		return v('div', { classes: this.classes(css.inputWrapper) }, [
+			w(TextInput, {
+				describedBy: this._errorId,
+				disabled,
+				invalid,
+				label,
+				maxLength,
+				minLength,
+				name,
+				placeholder,
+				readOnly,
+				required,
+				type,
+				value,
+				onBlur,
+				onInput
+			}),
+			invalid === true ? v('span', {
+				id: this._errorId,
+				classes: this.classes(css.error),
+				'aria-live': 'polite'
+			}, [ errorMessage ]) : null
+		]);
+	}
+}
+```
+{% endsolution %}
+
+{% instruction 'Use `ValidatedTextInput` within `WorkerForm`' %}
+
+Now that `ValidatedTextInput` exists, let's swap it with `TextInput` in `WorkerForm`, and write some error message text while we're at it:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:75-115 %}
```

Comment:
Do we need to show the import for the `ValidatedTextInput` also?

Reply:
Also fails at this point if the user is following instructions step by step... because needs the validate functions... do we think this is okay... or should we at least put an aside in?